### PR TITLE
Add preflight-lib reference

### PR DESCRIPTION
**Pull request by sebrandon1**
Follow up to #592 

Building upon work done by @komish in branch: https://github.com/komish/openshift-preflight/tree/add-lib-ha

- Adds the `preflight` test suite.
- Brings in external dependency from openshift-preflight repo.

### CODE REVIEW

File: pkg/provider/operators.go [hunk 1]
```diff
@@ -33,12 +38,32 @@ type Operator struct {
...
 }
 
 func (op *Operator) String() string {
 	return fmt.Sprintf("csv: %s ns:%s subscription:%s", op.Name, op.Namespace, op.SubscriptionName)
 }
 
+func (op *Operator) SetPreflightResults() error {
+	bundleImage := op.InstallPlans[0].BundleImage
+	indexImage := op.InstallPlans[0].IndexImage
+	check := plib.NewOperatorCheck(bundleImage, "/Users/bpalm/.kube/config", indexImage)
```
Comment (greyerof) [inline pkg/provider/operators.go:51]: We should add a way to retrieve the kubeconfig file used by the clientsholder so you can use it here.

File: cnf-certification-test/preflight/suite.go [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+	ginkgo.ReportAfterEach(results.RecordResult)
+
+	testPreflightContainers(&env)
+	testPreflightOperators(&env)
+})
+
+func testPreflightOperators(env *provider.TestEnvironment) {
+	// Loop through all of the operators, run preflight, and set their results into their respective object
+	for _, op := range env.Operators {
+		err := op.SetPreflightResults()
```
Comment (greyerof) [inline cnf-certification-test/preflight/suite.go:47]: A couple of comments here:
- `op.RunPreflightChecks()` or `op.RunPreflightTestCases` make ore sense to me here.
- There's an issue here: as this code will run as part of the Describe block (outside an `It` block), both operator and container preflight checks will run even if no focus/label for this preflight tests was provided.

File: pkg/provider/operators.go (hunk 1 above)
Comment (sebrandon1) [reply pkg/provider/operators.go:51]: I'm currently open to suggestions about how to gather that info.  😄 I wasn't able to find anything straight forward about how to retrieve a KUBECONFIG from the client.

File: cnf-certification-test/preflight/suite.go (hunk 2 above)
Comment (sebrandon1) [reply cnf-certification-test/preflight/suite.go:47]: This is a chicken/egg problem as we need the tests to be ran to be able to dynamically create the It blocks.

File: pkg/provider/operators.go (hunk 1 above)
Comment (sebrandon1) [reply pkg/provider/operators.go:62]: Feature added in #644. 🎉

**Review by greyerof — COMMENTED** [review]
I've seen in the log output from the preflight checks that they're using tmp folders to create some files... we might have the same issue as in #644:
image

File: cnf-certification-test/preflight/suite.go [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+			}
+		}
+		testhelper.AddTestResultLog("Non-compliant", failedContainers, tnf.ClaimFilePrintf, ginkgo.Fail)
+		testhelper.AddTestResultLog("Error", erroredContainers, tnf.ClaimFilePrintf, ginkgo.Fail)
+	})
+}
+
+func gatherTestNamesFromContainerResults(containers []*provider.Container) map[string]plibRuntime.Result {
+	testEntries := make(map[string]plibRuntime.Result)
+	for _, cut := range containers {
```
[the remark this answers is not in the document]
Comment (sebrandon1) [reply cnf-certification-test/preflight/suite.go:164]: Just for record-keeping, I'm just using this function to collect all of the possible test names from the results structs we have gathered and that way building the Ginkgo It blocks.

**Comment by sebrandon1** [comment]
Removing the proof of concept label.  The refactor PR merged today and I have updated this PR to point at it.

https://github.com/redhat-openshift-ecosystem/openshift-preflight/pull/825

This PR above merged into an upstream branch called `lib` but is still accessible via a go mod replace directive.

**Comment by sebrandon1** [comment]
I'm currently checking through these failures:

```
Summarizing 4 Failures:
  [FAIL] preflight [It] preflight-HasUniqueTag [common, preflight, preflight-HasUniqueTag]
  /Users/bpalm/Repositories/go/src/github.com/test-network-function/cnf-certification-test/pkg/testhelper/testhelper.go:82
  [FAIL] preflight [It] preflight-RunAsNonRoot [common, preflight, preflight-RunAsNonRoot]
  /Users/bpalm/Repositories/go/src/github.com/test-network-function/cnf-certification-test/pkg/testhelper/testhelper.go:82
  [FAIL] preflight [It] preflight-HasLicense [common, preflight, preflight-HasLicense]
  /Users/bpalm/Repositories/go/src/github.com/test-network-function/cnf-certification-test/pkg/testhelper/testhelper.go:82
  [FAIL] preflight [It] preflight-DeployableByOLM [common, preflight, preflight-DeployableByOLM]
  /Users/bpalm/Repositories/go/src/github.com/test-network-function/cnf-certification-test/pkg/testhelper/testhelper.go:82
  ```

**Comment by sebrandon1** [comment]
@ramperher I think the image credentials thing is the last thing this PR needs to pass before it can merge.  Do we have to change something on either our side or the DCI side to get it to be able to pull images correctly?

**Comment by ramperher** [comment]
> @ramperher I think the image credentials thing is the last thing this PR needs to pass before it can merge. Do we have to change something on either our side or the DCI side to get it to be able to pull images correctly?

I think the only missing thing is to provide to tnf container the configuration that preflight needs to retrieve the credentials to access to the targeted registries. This has to be modified in tnf side; a way of providing some files and environment variables to achieve this. And the files and values for the environment variables can be provided from our side (and, in the end, for whatever user of this repo).

In particular, we need two files that have to be copied in the tnf container in a well known location:

- One file containing the pullsecrets to access to the targeted registries (let's call it config.yaml)
- Another file containing the preflight configuration, where it is included the reference to the pullsecrets file path (let's call it creds.json). Here's an example of a [template](https://github.com/redhat-cip/dci-openshift-app-agent/blob/master/roles/preflight/templates/config_container.yaml.j2#L2) we use on DCI to generate config.yaml, where dockerConfig variable refers to the file containing the pullsecrets.

And also, we need to export `OPENSHIFT_AUTH` environment variable within tnf referencing the pullsecrets file path (path/to/creds.json within the tnf container).

Then, preflight parses this config file with [Viper](https://github.com/redhat-openshift-ecosystem/openshift-preflight/blob/21868850ba67a0348f43cb73649d662ea60bab1f/internal/runtime/config_test.go#L25), and then use this [docker credentials](https://github.com/redhat-openshift-ecosystem/openshift-preflight/blob/21868850ba67a0348f43cb73649d662ea60bab1f/internal/engine/engine.go#L94) while using crane to pull the image

So, tldr, we need:

- tnf must accept two files in a well known file path within it: config.yaml and creds.json.
- tnf has to export OPENSHIFT_AUTH environment variable pointing to path/to/creds.json.
- Preflight must be run from the file path where creds.json is saved

Maybe we can have a sync to review this tomorrow if you want, and then clarify what's needed.

**Comment by sebrandon1** [comment]
Interesting, we might have to generate some sort of functional ID or something to generate credentials to pass into the image.  Store them in Github and then pass them through when we build.

Yes, let's discuss tomorrow.

**Comment by ramperher** [comment]
@sebrandon1 , by the way, regarding this change, because I've seen you've submitted some updates. When you feel this is working in your side with the constraints we commented regarding the credentials to pull images from registries, so that this is ready for review on DCI side again, just ping me, because I'll need to do some changes on DCI to also propagate the credentials to be used. In the meanwhile, DCI job will continue failing.

**Comment by sebrandon1** [comment]
@ramperher I'm trying to stick to the preflight team's environment variable `PFLT_DOCKERCONFIG` found [here](https://github.com/redhat-openshift-ecosystem/openshift-preflight/blob/main/docs/CONFIG.md).  If this is set (or if there is a `"$HOME/.docker/config.json"` set, it will work.  (Fingers crossed). 😄

**Comment by ramperher** [comment]
@sebrandon1 , despite mounting correctly the config.json file:

```
Mounting 1 dockercfg volume(s):
-v /tmp/ansible.2isc5w7h/config.json:/usr/tnf/dockercfg/config:Z
```

Preflight is struggling with authentication issues:

```
[37mDEBUG  [0m[Jan 25 14:37:41.220][suite.go: 32] Entering preflight suite                     
[37mDEBUG  [0m[Jan 25 14:37:41.220][suite.go: 76] Running preflight container tests for: test  
[31mERROR  [0m[Jan 25 14:37:41.234][containers.go: 88] failed to pull remote container: Get "https://registry.dfwt5g.lab:4443/v2/": x509: certificate signed by unknown authority 
[31mFATAL  [0m[Jan 25 14:37:41.234][suite.go: 79] failed running preflight on image: registry.dfwt5g.lab:4443/testnetworkfunction/cnf-test-partner:latest error: failed to pull remote container: Get "https://registry.dfwt5g.lab:4443/v2/": x509: certificate signed by unknown authority 
```

Is there any way to see if preflight is taking correctly the config file?

**Review by ramperher — APPROVED** [review]
you have my :+1: from DCI perspective!

**Comment by sebrandon1** [comment]
For reviewers, do we have a preference if these preflight tests should be tagged `TagCommon`?  What about how we classify them as telco/non-telco/etc @mwlinca?

**Comment by ramperher** [comment]
> For reviewers, do we have a preference if these preflight tests should be tagged `TagCommon`? What about how we classify them as telco/non-telco/etc @mwlinca

From my point of view, until having an official resolution about how tnf and preflight will be tested, I would use a separate label for preflight, not including it within the `common` label

**Review by shimritproj — APPROVED** [review]

File: cnf-certification-test/preflight/suite.go [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+		// Store the test names into the Catalog map for results to be dynamically printed
+		aID := identifiers.AddCatalogEntry(testName, common.PreflightTestKey, testEntry.Metadata().Description, testEntry.Help().Suggestion, "", "", "", "", false, identifiers.TagCommon)
+		testID, tags := identifiers.GetGinkgoTestIDAndLabels(aID)
+
+		logrus.Infof("Testing ginkgo test: %s ID: %s", testName, testID)
+		GeneratePreflightContainerGinkgoTest(testName, testID, tags, env.Containers)
+	}
+}
+
+func GeneratePreflightContainerGinkgoTest(testName, testID string, tags []string, containers []*provider.Container) {
```
Comment (jmontesi) [inline cnf-certification-test/preflight/suite.go:95]: Should you make this function private?

File: cnf-certification-test/preflight/suite.go [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+	// Based on a single test "name", we will be passing/failing in Ginkgo.
+	// Brute force-ish type of method.
+	ginkgo.It(testID, ginkgo.Label(tags...), func() {
+		// Collect all of the failed and errored containers
+		var failedContainers []string
+		var erroredContainers []string
+		for _, cut := range containers {
+			for _, i := range cut.PreflightResults.Passed {
```
Comment (jmontesi) [inline cnf-certification-test/preflight/suite.go:103]: See comment above about the name `i`.

File: cnf-certification-test/preflight/suite.go [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+	// Loop through all of the containers, run preflight, and set their results into their respective object
+	for _, cut := range env.Containers {
+		logrus.Debugf("Running preflight container tests for: %s", cut.Name)
+		err := cut.SetPreflightResults(preflightImageCache, env.IsPreflightInsecureAllowed())
+		if err != nil {
+			logrus.Fatalf("failed running preflight on image: %s error: %v", cut.Image, err)
+		}
+	}
+
+	containerTestEntries := gatherTestNamesFromContainerResults(env.Containers)
```
Comment (jmontesi) [inline cnf-certification-test/preflight/suite.go:82]: Isn't this map the same as the one in `preflightImageCache`?

File: pkg/provider/operators.go [hunk 7]
```diff
@@ -57,6 +64,48 @@ func (op *Operator) String() string {
...
+	artifactsWriter, err := artifacts.NewMapWriter()
+	if err != nil {
+		return err
+	}
+	ctx := artifacts.ContextWithWriter(context.Background(), artifactsWriter)
+
+	opts := []plibOperator.Option{}
+
+	// Check to make sure that the environment variable is set
+	if val := os.Getenv("PFLT_DOCKERCONFIG"); len(val) > 0 {
```
Comment (jmontesi) [inline pkg/provider/operators.go:82]: See the other comment about environment variables.

File: cnf-certification-test/preflight/suite.go [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+func GeneratePreflightOperatorGinkgoTest(testName, testID string, tags []string, operators []*provider.Operator) {
+	// Based on a single test "name", we will be passing/failing in Ginkgo.
+	// Brute force-ish type of method.
+	ginkgo.It(testID, ginkgo.Label(tags...), func() {
+		// Collect all of the failed and errored containers
+		var failedOperators []string
+		var erroredOperators []string
+		for _, op := range operators {
+			for _, i := range op.PreflightResults.Passed {
+				if i.Name() == testName {
```
Comment (jmontesi) [inline cnf-certification-test/preflight/suite.go:137]: See comment about the name `i`.

File: cnf-certification-test/preflight/suite.go [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+	"github.com/test-network-function/cnf-certification-test/pkg/provider"
+	"github.com/test-network-function/cnf-certification-test/pkg/testhelper"
+	"github.com/test-network-function/cnf-certification-test/pkg/tnf"
+)
+
+var _ = ginkgo.Describe(common.PreflightTestKey, func() {
+	logrus.Debugf("Entering %s suite", common.PreflightTestKey)
+	env := provider.GetTestEnvironment()
+	ginkgo.BeforeEach(func() {
+		env = provider.GetTestEnvironment()
```
[the remark this answers is not in the document]
Comment (sebrandon1) [reply cnf-certification-test/preflight/suite.go:35]: There _was_ a reason I did this on purpose, but I can't recall at the moment what that reason was.  😅 

I'll retry today and see if removing it works.

[the remark this answers is not in the document]
Comment (sebrandon1) [reply cnf-certification-test/preflight/suite.go:35]: I _think_ it's because the `It` blocks aren't created until after the tests are ran so I needed to "prime-the-pump" sort of thing.

File: pkg/provider/containers.go [hunk 10]
```diff
@@ -62,6 +69,47 @@ func (c *Container) GetUID() (string, error) {
 	return uid, nil
 }
 
+func (c *Container) SetPreflightResults(preflightImageCache map[string]plibRuntime.Results, allowInsecure bool) error {
+	var results plibRuntime.Results
+	if _, exists := preflightImageCache[c.Image]; !exists {
```
[the remark this answers is not in the document]
Comment (sebrandon1) [reply pkg/provider/containers.go:74]: Good idea, I can fix.

File: pkg/provider/containers.go [hunk 11]
```diff
@@ -62,6 +69,47 @@ func (c *Container) GetUID() (string, error) {
...
+		opts := []plibContainer.Option{}
+
+		// Check to make sure that the environment variable is set
+		if val := os.Getenv("PFLT_DOCKERCONFIG"); len(val) > 0 {
```
[the remark this answers is not in the document]
Comment (sebrandon1) [reply pkg/provider/containers.go:78]: Good idea, I'll move it to there for `envconfig` to handle.

File: cnf-certification-test/preflight/suite.go (hunk 4 above)
Comment (sebrandon1) [reply cnf-certification-test/preflight/suite.go:95]: Done.

File: cnf-certification-test/preflight/suite.go [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+		}
+		testhelper.AddTestResultLog("Non-compliant", failedOperators, tnf.ClaimFilePrintf, ginkgo.Fail)
+		testhelper.AddTestResultLog("Error", erroredOperators, tnf.ClaimFilePrintf, ginkgo.Fail)
+	})
+}
+
+func gatherTestNamesFromContainerResults(containers []*provider.Container) map[string]plibRuntime.Result {
+	testEntries := make(map[string]plibRuntime.Result)
+	for _, cut := range containers {
+		for _, i := range cut.PreflightResults.Passed {
```
[the remark this answers is not in the document]
Comment (sebrandon1) [reply cnf-certification-test/preflight/suite.go:164]: I think I will switch these to `r` for result.

File: cnf-certification-test/preflight/suite.go [hunk 13]
```diff
@@ -0,0 +1,194 @@
...
+					erroredOperators = append(erroredOperators, op.String())
+				}
+			}
+		}
+		testhelper.AddTestResultLog("Non-compliant", failedOperators, tnf.ClaimFilePrintf, ginkgo.Fail)
+		testhelper.AddTestResultLog("Error", erroredOperators, tnf.ClaimFilePrintf, ginkgo.Fail)
+	})
+}
+
+func gatherTestNamesFromContainerResults(containers []*provider.Container) map[string]plibRuntime.Result {
```
[the remark this answers is not in the document]
Comment (sebrandon1) [reply cnf-certification-test/preflight/suite.go:161]: Changed to `getUniqueTestEntriesFromContainerResults`.

File: cnf-certification-test/preflight/suite.go (hunk 8 above)
Comment (sebrandon1) [reply cnf-certification-test/preflight/suite.go]: Done.

File: pkg/provider/operators.go (hunk 7 above)
Comment (sebrandon1) [reply pkg/provider/operators.go]: Done.  Moved to the envconfig parameter struct.

File: cnf-certification-test/preflight/suite.go (hunk 5 above)
Comment (sebrandon1) [reply cnf-certification-test/preflight/suite.go]: Done.

**Review by jmontesi — APPROVED** [review]

File: pkg/provider/containers.go [hunk 14]
```diff
@@ -62,6 +69,51 @@ func (c *Container) GetUID() (string, error) {
...
+	}
+
+	check := plibContainer.NewCheck(c.Image, opts...)
+
+	// Create artifacts handler
+	artifactsWriter, err := artifacts.NewMapWriter()
+	if err != nil {
+		return err
+	}
+	ctx := artifacts.ContextWithWriter(context.Background(), artifactsWriter)
```
Comment (edcdavid) [inline pkg/provider/containers.go:101]: Configure a logger  to retrieve the logs from preflights. I tried to run it but it only displayed the info logs (V(0)), not sure how to set the level (see https://github.com/redhat-openshift-ecosystem/openshift-preflight/blob/main/docs/LIBRARY.md ):
		
// [2] Configure a logr.Logger compliant logger of your choice, and add it to the context
		logbytes := bytes.NewBuffer([]byte{})
		checklogger := log.Default()
		checklogger.SetOutput(logbytes)
		logger := stdr.New(checklogger)
		ctx = logr.NewContext(ctx, logger.V(3))

	// [4] Accessing logs
	fmt.Println("The preflight log contained:")
	fmt.Println(logbytes.String())

Comment (sebrandon1) [reply pkg/provider/containers.go:101]: Good find, I didn't know this was exposed.  I'll play with adding this.

Comment (sebrandon1) [reply pkg/provider/containers.go:101]: DEBUG  [Feb 14 11:46:48.910][suite.go: 73] Running preflight container tests for: jack  
INFO   [Feb 14 11:46:48.910][containers.go: 77] Running preflight container test against image: quay.io/testnetworkfunction/cnf-test-partner:latest with name: jack 
INFO   [Feb 14 11:47:32.729][containers.go: 125] 2023/02/14 11:47:18 "level"=0 "msg"="check completed" "result"="PASSED" "check"="HasLicense"
2023/02/14 11:47:18 "level"=0 "msg"="check completed" "result"="PASSED" "check"="HasUniqueTag"
2023/02/14 11:47:18 "level"=0 "msg"="check completed" "result"="PASSED" "check"="LayerCountAcceptable"
2023/02/14 11:47:18 "level"=0 "msg"="check completed" "result"="PASSED" "check"="HasNoProhibitedPackages"
2023/02/14 11:47:18 "level"=0 "msg"="check completed" "result"="PASSED" "check"="HasRequiredLabel"
2023/02/14 11:47:18 "level"=0 "msg"="USER tnf-user specified that is non-root"
2023/02/14 11:47:18 "level"=0 "msg"="check completed" "result"="PASSED" "check"="RunAsNonRoot"
2023/02/14 11:47:29 "level"=0 "msg"="check completed" "result"="PASSED" "check"="HasModifiedFiles"
2023/02/14 11:47:30 "level"=0 "msg"="check completed" "result"="PASSED" "check"="BasedOnUbi"
2023/02/14 11:47:30 "level"=0 "msg"="This image's tag latest will be paired with digest sha256:61d60fe79642048e8b69e318fbc1616e3c0cbd792aa5573d327b950b0f9e1926 once this image has been published in accordance with Red Hat Certification policy. You may then add or remove any supplemental tags through your Red Hat Connect portal as you see fit."

Comment (sebrandon1) [reply pkg/provider/containers.go:101]: Yeah I'm only seeing what is essentially the bare minimum of logs.  I wonder if in the future we might be able to specify a more verbose log level.

Comment (sebrandon1) [reply pkg/provider/containers.go:108]: See logging lines added in https://github.com/test-network-function/cnf-certification-test/pull/631/commits/f644eddca80dafc0effcb345820eee4f8aaf3c40

**Review by edcdavid — APPROVED** [review]
